#3180. Add tests for `FunctionToJSExportedDartFunction`. Part 2.

sgrekhov opened the pull request:
Please note that I use `Expect.equals(globalContext, pThis);` to check the value of `this`. Is it Ok to make an assumption that `this == globalContext`?

eernstg reviewed (commented):
Added a few comments. The main thing is that I'm surprised that `toJSCaptureThis` is documented as being identical to `toJS` except for capturing and passing `this`, but the requirements on the types used in the signature of the receiver seem to be different (it looks like nothing is required for `pThis`).

sgrekhov commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t03.dart:
```diff
@@ -0,0 +1,48 @@
+/// parameter at the beginning of the parameter list to handle this.
+///
+/// @description Check that if this getter is called on a torn off constructor
+/// it is a run-time error to call it in JavaScript.
```
Yes, thank you. Forgot to updated the description. Fixed.

sgrekhov commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart:
```diff
@@ -0,0 +1,42 @@
+/// Any [Function] that is converted with this member should take in an extra
+/// parameter at the beginning of the parameter list to handle this.
+///
+/// @description Check that this property returns JavaScript function that wraps
```
Fixed everywhere.

sgrekhov commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart:
```diff
@@ -0,0 +1,42 @@
+import '../../../Utils/expect.dart';
+import '../js_utils.dart';
+
+int foo(pThis) {
```
Awaiting for the clarification.

srujzs replied:
Sergey filed a few issues, but the central problem is that I never duplicated the static error checks for `toJSCaptureThis`: https://github.com/dart-lang/sdk/issues/61439.

sgrekhov reviewed (commented):
Updated but waitig for further info.

srujzs commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart:
```diff
@@ -0,0 +1,42 @@
+import '../js_utils.dart';
+
+int foo(pThis) {
+  Expect.equals(globalContext, pThis);
```
Once `pThis` is updated to e.g. `JSAny`, the warnings around equality and identicality are the same as other instances of JS interop. Equality will check the underlying JS values, but identicality will check the wrappers and therefore fail in dart2wasm.

srujzs commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:
```diff
@@ -0,0 +1,50 @@
+    return a.id;
+  }
+  static String bar(pThis, String v) {
+    Expect.equals(globalContext, pThis);
```
Another alternative is using `callAsFunction` (or the JS equivalent: `call`) which allows you to pass a `this` value explicitly.

eernstg replied:
@sgrekhov, I think this can be settled in a separate PR.

sgrekhov replied:
Yes, I'll add these tests in a separate PR

sgrekhov replied:
> Another alternative is using `callAsFunction` (or the JS equivalent: call) which allows you to pass a this value explicitly.

Ups! There are no tests for `callAsFunction`. I'll add them too.

sgrekhov replied:
> Another alternative is using `callAsFunction` (or the JS equivalent: `call`) which allows you to pass a `this` value explicitly.

@srujzs I'm trying to check value of `this` via `callAsFunction`. The code below works on `dart2was` but in case of `dart2js` it produces `TypeError: Instance of 'String': type 'UnknownJavaScriptObject' is not a subtype of type 'String'`. Am I doing anything wronr? Or this is an issue?

```dart
void foo(JSObject pThis) {
  Expect.equals("Value of this", (pThis as JSString).toDart); // Error here on dart2js
}

main() {
  globalContext["jsFoo"] = foo.toJSCaptureThis;
  JSFunction jsFoo = globalContext["jsFoo"] as JSFunction;
  jsFoo.callAsFunction("Value of this".toJS);
}
```

srujzs replied:
That's because `pThis` isn't a `JSObject` under the hood :D

You likely want `void foo(String pThis)`. It works in dart2wasm because `JSObject` and `JSString` are the same box. In dart2js and DDC, we have stronger type-checking.

srujzs commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t03.dart:
```diff
@@ -0,0 +1,48 @@
+    Expect.equals(globalContext, pThis);
+  }
+
+  @JS()
```
Note that `@JS` does nothing on constructors here and below, regardless of `external` or not. It's still on my plate to add warnings though.

eernstg reviewed (commented):
Very good, I think this one is now subject to a few changes, and there may be another PR with some other changes. So I'll just await further input on this PR for now.

sgrekhov commented:
Thank you! Updated.
- Removed `@JS()` from the constructors (otherwise the tests fail once these annotations start producing warnings).
- Changed the type of the `pThis` parameter to `JSObject` (again, otherwise the tests will fail when type checking is added). Please let me know if this type should instead be `JSAny`.
- Added a test checking arguments of types `JSObject` and `JSAny`.

Please take another look.

srujzs commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:
```diff
@@ -20,15 +20,17 @@ import 'dart:js_interop';
+C? foo(JSObject pThis) {return null;}
 
-void bar(pThis, C c) {}
+void bar(JSObject pThis, C c) {}
```
nit: It may be worth having a function with all valid return and parameter types except for the first parameter (the `this` value) to check that we still catch that.

sgrekhov replied:
There is `void corge(pThis) {}` on the line 33.

srujzs replied:
Ah I missed that, thanks!

srujzs commented:
> Please let me know if this type should instead be JSAny.

It can be any valid type for interop members. `JSObject` or `JSAny` works in this case because `globalThis` is a JS object, but if we add tests to pass a symbol, `null`, etc. for `this`, then we'd naturally want to change the type we write.

eernstg commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t04.dart:
```diff
@@ -0,0 +1,54 @@
+}
+
+JSString bar(
+  pThis,
```
Probably:
```suggestion
  JSObject pThis,
```

eernstg commented on LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:
```diff
@@ -0,0 +1,60 @@
+
+void quux(JSObject pThis, Symbol s) {}
+
+void corge(pThis) {}
```
(This one was already mentioned somewhere)
```suggestion
void corge(JSObject pThis) {}
```

eernstg replied:
Oh, of course, there error may seem to be here, but the error message is at the invocation of `toJSCaptureThis`.

eernstg reviewed (approved):
Looks good, but it looks like one parameter type is still missing in a couple of places. Otherwise I think it's ready to land.
